### PR TITLE
feat(guardrails): add input_messages to scan only the latest turn

### PR DESCRIPTION
**Pull request by nic-6443**
## Problem

IDE and agent clients replay the whole conversation on every call. An input guardrail therefore re-reads every earlier message on every request, so a rule that matched one message keeps refusing the rest of the session even though the new prompt is clean. A customer's regex rule is stuck this way today: the only way out is to start a new conversation or drop the rule.

## What this adds

A new optional top-level field on the guardrail resource, common to all twelve kinds:

```yaml
input_messages: all | latest_turn   # optional, default "all"
```

- **`all`** — today's behaviour, unchanged. Every message in the request is read: system, user, assistant, tool.
- **`latest_turn`** — only the messages after the last assistant message, with system messages excluded. A *trailing* assistant message does not count: that is a prefill, text the caller wrote for the model to continue rather than a turn the model has answered, so it belongs to the current turn and is scanned with it. "Trailing" is measured against the last non-system message, since system messages sit outside the window wherever they are — an assistant message followed only by system ones has still answered nothing. In practice that is this turn's user message plus the tool results answering it (`role: "tool"` on the chat wire, `tool_result` blocks on `/v1/messages`, `function_call_output` items on `/v1/responses`). A request with no assistant message narrows to every non-system message. Messages the model has already replied to are neither screened nor rewritten.

The field is inert at the output hook, which always reads the whole response.

**The default is unchanged**, and a row that does not carry the field behaves exactly as before.

## How it is wired

The narrowing governs **both** input passes, and the window is per row — a chain that mixes an `all` row with a `latest_turn` row gives each what it asked for.

- **Block/allow check.** Every endpoint already converts its request into one `ChatFormat` and calls the same trait method, so the narrowing lives in the chain fold: each member is handed either the request or a narrowed view of it. No endpoint-side change, and the single-input surfaces (embeddings, rerank, images, audio, MCP, passthrough, …) are a natural no-op.
- **Masking.** The mask passes rewrite raw wire slots and have no parsed view to index against, so each input walker in `aisix-proxy::redact` answers the same boundary question for its own shape and tags every slot it offers. The chain then hands a narrowed row only the in-window slots and splices its reply back onto the positions it was given; a row on `all` still sees and rewrites everything. The per-kind segment hooks are untouched — a kind never learns its own window.

The boundary is therefore stated twice, once per representation. Both halves are pinned per protocol in the e2e suite, deliberately, because a drift between them is the failure this design can have.

## Behaviour changes

**Masking under `latest_turn` does not rewrite history.** Messages outside the window reach the upstream exactly as the caller sent them. A rule whose job is to mask the whole conversation belongs on `all`. This is stated in the field description too.

**`/v1/responses` now scans a replayed tool call's `name` and `arguments`, under `all` as well.** That API spells a model turn as a typed item carrying no `role`, so the input translator read an agent's whole tool loop as user text — and dropped `function_call` items entirely, because they carry none of the keys the scan used to read. A block rule that fires on `/v1/chat/completions`, where the same replayed call rides `extra["tool_calls"]` and has always been scanned, was bypassable by moving the payload into a tool call on this surface. The mask walker already rewrote both keys, so the two passes now read the same slots.

The item mapping is now:

| `input[]` item | role |
| --- | --- |
| `role: "assistant"` message | assistant |
| `function_call`, `custom_tool_call` | assistant (`name` + `arguments`/`input` scanned) |
| `reasoning` | assistant |
| `function_call_output`, `custom_tool_call_output` | tool |
| `role: "system"` / `"developer"`, `instructions` | system |
| everything else | user |

An operator running a block rule on `/v1/responses` whose pattern appears in a replayed tool call's arguments will start seeing refusals that previously passed. That is the gap being closed, not a regression.

**Two kinds stop screening replayed tool results on `/v1/responses`.** `semantic` and `azure_content_safety_text_moderation` read only user-role messages under their default `text_source` (`user_messages` / `concatenate_user_content`). A `function_call_output` used to be mislabelled as a user message, so they happened to scan it; now that it is correctly a tool result, they do not — exactly as a `role: "tool"` message already sits outside what they read on `/v1/chat/completions`. This is the two surfaces agreeing rather than a new gap, but it is a coverage reduction for an existing rule with unchanged configuration, so: an operator who wants tool results screened by either kind sets `text_source` to its all-messages value, on either surface. A unit test names the behaviour so it is not rediscovered as a bug.

The same translator also stops dropping any item that carries no readable text. It used to drop them, which removed the very boundary the window opens after: a `reasoning` item whose only payload is `encrypted_content` has nothing to scan, so the check pass saw no model turn and widened back to the whole conversation while the mask walkers correctly treated the messages before it as history. Agent clients replay that shape on every turn when reasoning summaries are off. The parsed view is now one message per `input[]` item whatever the item carries, which is what keeps it aligned with the wire the walkers read; an empty message contributes nothing to any kind's scan.

`redact_responses_item` also gained the custom-tool arms — `custom_tool_call`'s `input` and `custom_tool_call_output`'s `output` — which the scan reads and the mask walk did not, so a Mask rule reported a hit there and forwarded the match unmasked. A tool `name` is still scanned and never rewritten, matching `/v1/chat/completions`, where a tool-call name is treated as structural.

## Tests

`tests/e2e/src/cases/guardrail-input-messages-latest-turn-e2e.test.ts` (29 cases) drives a real gateway + etcd + mock upstream. For each of `/v1/chat/completions`, `/v1/messages` and `/v1/responses`, four boundary cases — marker in replayed history, in the current user message, in this turn's tool result, in the system prompt — run twice: once with a `keyword` row (which answers from the check pass) and once with a `custom` script row (which answers from the segment pass), so each half of the boundary rule is pinned independently. Plus three `/v1/responses` cases for the typed model items — including one where the model's turn carries no readable text and both halves must still agree — and two mask cases asserting the mock upstream received the history byte-identical while the current turn was rewritten — one per channel, since the segment channel narrows by splicing a slot subset and the sync channel (`kind: pii`) narrows by filtering chain members.

Unit tests additionally pin the trailing-prefill rule on both halves, the mixed-window mask composition (one row on `all`, one on `latest_turn`, rewriting the same slot list), and the Responses role mapping.

Each half was mutation-checked: disabling the chain narrowing reds exactly the seven check-pass cases, disabling the wire-level windowing reds exactly the seven segment/mask cases, and dropping `name`/`arguments` from the Responses scan reds exactly the tool-call case. Unit tests pin the two boundary functions and the Responses role mapping directly. The existing `guardrail-keyword-message-locations-e2e` suite, which asserts whole-history scanning, stays green under the default.

## Control plane

This is a user-configurable surface, so it is not reachable until the control plane exposes it: `cp-admin.yaml` schema + Go model/validation/projection, the dashboard form, and CP↔DP integration tests. Tracked on api7/AISIX-Cloud#1558.

Two notes for that side. The control plane should reject `hook_point: output` together with `input_messages: latest_turn` — the gateway accepts the pair and simply ignores the field there, so nothing breaks, but it is a configuration the operator did not mean. And this is a new optional field rather than a new value on an existing one, so a gateway older than this release ignores it and keeps scanning the whole conversation: the feature is unavailable there, never fail-open, and needs no compatibility gate.

Fixes api7/AISIX-Cloud#1558

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/aisix-guardrails/src/lib.rs [hunk 1]
```diff
@@ -810,6 +840,23 @@ pub trait Guardrail: Send + Sync + 'static {
...
+    /// configured `input_messages: latest_turn` for out-of-window text, so
+    /// a mask rule on that setting rewrites the current turn and leaves the
+    /// replayed history byte-identical. A leaf guardrail has no window of
+    /// its own — the setting belongs to the chain member, not the kind —
+    /// so the default ignores the flag and every caller that has no window
+    /// to report (the whole response side, and the single-input endpoints)
+    /// keeps using [`Self::redact_input_text`] directly.
+    fn redact_input_text_in_turn(&self, text: &str, in_latest_turn: bool) -> Option<Redaction> {
+        let _ = in_latest_turn;
+        self.redact_input_text(text)
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply crates/aisix-guardrails/src/lib.rs:877]: Right about `redact_input_text_in_turn` — fixed in 2833291a, it now delegates to `self.current()` and forwards the flag.

Two clarifications on the rest. There is no output-side window hook to forward: `input_messages` governs the input hook only and is inert on the output hook by design, so `redact_output_text` has no windowed variant. And the other window-aware hook, `moderate_input_segments_in_turn`, is deliberately left alone: this wrapper does not forward `moderates_segments` either (that predates this PR), so it always reports `false` and no call site can reach the segment pass through it. Forwarding just the windowed half would be misleading rather than safer. Noted in the doc comment on the new method.

Worth saying the wrapper is unreachable from the proxy in any case — everything resolves chains through `LiveGuardrailIndex`. It is public, which is why the one-line omission was still worth closing.

File: crates/aisix-proxy/src/responses.rs [hunk 2]
```diff
@@ -996,11 +996,11 @@ fn responses_input_to_chat(model: &str, body: &Value) -> ChatFormat {
...
-                let role = item.get("role").and_then(|v| v.as_str()).unwrap_or("user");
-                messages.push(match role {
-                    "assistant" => ChatMessage::assistant(text),
-                    "system" | "developer" => ChatMessage::system(text),
-                    _ => ChatMessage::user(text),
+                messages.push(match responses_item_role(item) {
+                    Role::Assistant => ChatMessage::assistant(text),
+                    Role::System => ChatMessage::system(text),
+                    Role::Tool => ChatMessage::tool(text),
+                    Role::User => ChatMessage::user(text),
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply crates/aisix-proxy/src/responses.rs:1014]: Good catch, and it was real on the commit you reviewed (2423c529). Fixed in fc502d23.

The root cause was `responses_input_to_chat` dropping items whose text came back empty, which is exactly what removed the boundary the redact pass could still see. It now emits one message per `input[]` item whatever the item carries, so the parsed view is aligned one-to-one with the wire and both passes index the same list. An empty message contributes nothing to any kind`s scan, so keeping it costs nothing.

The two boundary predicates in `redact.rs` also stopped re-deriving the answer — `responses_item_is_assistant` and `responses_item_is_system` now call `responses::responses_item_role`, so the protocol has one role mapping rather than two. That closed a second divergence on `{"role": "user", "type": "reasoning"}`, which the old predicate treated as a boundary and the translator did not.

There is a third boundary change in the same commit that your scenario touches: a *trailing* assistant-side item no longer closes the window. It used to, which left the window empty and turned an appended assistant message into a one-line bypass of every `latest_turn` rule.

Pinned by unit tests on both halves per protocol (`redact::tests::a_trailing_assistant_message_does_not_close_the_window_on_any_wire`, `responses::tests::responses_input_keeps_one_message_per_item_even_when_text_is_empty`) and by the e2e case `"/v1/responses: a model turn carrying no readable text still opens the window"`, which drives the keyword row (check pass) and the custom-script row (segment pass) over the same body and requires both to agree.

File: crates/aisix-guardrails/src/lib.rs [hunk 3]
```diff
@@ -164,6 +164,44 @@ pub(crate) fn message_scan_text(m: &ChatMessage) -> String {
...
+/// and a `latest_turn` rule goes quiet. Anthropic's documented
+/// assistant-prefill feature reaches the same shape by accident.
+///
+/// This is the CHECK pass's half of the rule. The masking walkers in
+/// `aisix-proxy::redact` apply the same rule to each wire shape directly,
+/// because their slots are raw JSON with no `ChatFormat` to index against;
+/// the e2e cases pin both halves per protocol.
+pub fn latest_turn_view(req: &ChatFormat) -> ChatFormat {
+    let answered = req.messages.len().saturating_sub(1);
+    let start = req.messages[..answered]
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply crates/aisix-guardrails/src/lib.rs]: Correct, and reachable — fixed in 225b24c2.

The prefill exclusion was measured against the last message rather than the last non-system one, so appending a system message after the prefill made the prefill look answered again and left the window holding system messages alone. Since those are excluded, that is an empty window: the same bypass one step further out.

System messages sit outside the window wherever they are, so an assistant message followed only by system ones has still answered nothing. All four boundary computations now find the last non-system message first and look for the closing assistant turn before it — `latest_turn_view` plus the three walkers you listed.

Pinned on both halves, including the `role: "system"` entry inside `messages[]` that Claude Code sends and the `developer` spelling on `/v1/responses`: `chain::tests::a_system_message_after_a_prefill_does_not_reopen_the_bypass` and `redact::tests::a_system_message_after_a_prefill_does_not_close_the_window`. Both go red if the non-system filter is removed.

File: tests/e2e/src/cases/guardrail-input-messages-latest-turn-e2e.test.ts [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+      if (!etcdReachable || !app || !upstream) {
+        ctx.skip();
+        return;
+      }
+      const HISTORY_MAIL = "history.person@example.com";
+      const CURRENT_MAIL = "current.person@example.com";
+
+      await waitConfigPropagation(async () => {
+        const before = upstream!.receivedRequests.length;
+        await chat(piiLatest.model, [
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply tests/e2e/src/cases/guardrail-input-messages-latest-turn-e2e.test.ts]: Fair — added in the commit above. Both mask cases now assert a 200 from the caller before reading the upstream body, and both readiness probes gate on `res.ok`.

The reasoning I would put on it rather than "check every return value": a masking row never refuses, so any non-200 means the request took a path the case is not describing, and the upstream body would then be evidence about a different request than the one the assertion claims. That is what makes the check worth having here.

One note on the citation — the "every function return value must be checked" guideline is not in this repo (no `.coderabbit.yaml`, and it does not appear in `AGENTS.md` / `CLAUDE.md`). Checking `res.ok` in the e2e suite is an established convention here, though, which is why the suggestion stands on its own.
